pkg/local-disk-manager/disk/controller.go

```diff
@@ -81,38 +81,36 @@ func (ctr *Controller) handleStaleDisks(existDisks []manager.Event) {
 		}
 
 		exist := false
-		// search by serial number
-		if ld.Spec.DiskAttributes.SerialNumber != "" {
-			for _, attr := range existDiskAttrs {
-				if ld.Spec.DiskAttributes.SerialNumber == attr.Serial {
-					log.WithField("serialNumber", attr.Serial).WithField("ldName", ld.Name).Info("Found existing disk serial number")
+		// get all devLinks of current ld
+		devLinkSet := make(map[string]struct{})
+		for _, devLink := range ld.Spec.DevLinks {
+			devLinkSet[devLink] = struct{}{}
+		}
+
+		// search by dev link
+		for _, attr := range existDiskAttrs {
+			existDevLinkSet := make(map[string]struct{})
+			for _, existDevLink := range attr.DevLinks {
+				existDevLinkSet[existDevLink] = struct{}{}
+			}
+			for key, _ := range devLinkSet {
+				// every key must exactly hit the existDiskAttrs
+				if _, ok := existDevLinkSet[key]; ok {
+					log.WithField("devLink", attr).WithField("ldName", ld.Name).Info("Found existing disk IDPath")
 					exist = true
+				} else {
+					// if one non-exist devLink was found, current attr will be ignored
+					log.WithField("devLink", attr).WithField("ldName", ld.Name).Info("Found a non-existing disk IDPath,")
+					exist = false
 					break
 				}
 			}
```

Comment:
`existDiskAttrs` contains all DevLinks of a speial disk e.g, IDPath, DevPath(pci-0000:03:00.0-scsi-0:0:10:0)

But the DevPath is unreliable, it can be different when node restarts. So, we use serial number to identify disk first, only use DevPath when serial number is missing.

Reply:
this may cause failure when serial number exist but devlinks changes after node restart.